nodeJS/express/express_forms_validation.md

```diff
@@ -0,0 +1,520 @@
+
+#### Installation
+
+We start as usual by installing the correct package in the *root* folder of our project.
+
+```bash
+npm install express-validator
+```
+
+To get started using the package, we first need to *require* it in our project
+
+```javascript
+const { body, validationResult } = require("express-validator");
+```
+
+`express-validator` comes with many functions for just about every form operation you could think of, but for this lesson, we'll be sticking with `body()` and `validationResult()`.
+
+### Understanding the body() function
+
+The `body()` function allows you to specify which fields in the request body should be validated and sanitized, as well as how to handle it.
+
+```javascript
+[
+  body("birthdate", "Must be a valid date.")
+          .optional({ values: "falsy" })
+          .isISO8601() // Enforce a YYYY-MM-DD format.
+];
+```
+
+This example marks `birthdate` field as optional, but still enforces the ISO8601 date format on inputs. This is because `{ values: "falsy }"` means values that aren't `undefined`, `null`, `false`, or `0` will still be validated.
+
+### Chaining validations
+
+You can also chain multiple validation methods, with unique error messages if the checks fail.
+
+```javascript
+[
+    body("name")
+    .trim()
+    .isLength({ min: 1 })
+    .withMessage("Name can not be empty.")
+    .isAlpha()
+    .withMessage("Name must only contain alphabet letters."),  
+];
+```
+
+This ensures that `name` is not only present and trimmed, but also only contains alphabet letters.
+
+While this might work for outputs we know won't have special characters, like names or ages, we also have to consider situations that do allow those characters. For example, when writing their "About Me" description, what would happen if the client decides to inject JavaScript code instead?
+
+```ejs
+<div>
+  About Me: <%- description %>
+</div>
+
+// The client then inputs the following as their page's About Me:
+<script>alert("Hacked!");</script>
+```
+
+When unescaped, this would be rendered into HTML as:
+
+```html
+<div>
+  About Me: <script>alert("Hacked!");</script>!
+</div>
+```
+
+To prevent this cross-site scripting (XSS) attack, we can *escape* the output (you may also see this referred to as *encoding*). Escaped HTML replaces special characters, like `<`, with their respective HTML entities, in this case `&lt;`. In EJS, we can escape the output using `<%= %>`.
+
+```ejs
+<div>
+  About Me: <%= username %>!
+</div>
+
+// The escaped output is now rendered harmless:
+// About Me: &lt;script&gt;alert(&quot;Hacked!&quot;);&lt;/script&gt;!
+```
+
+So why we don't chain `.escape()` in our Express server's `body()` function instead? We would then need to unescape the input when we go to output it again, or use unescaped output with `<%- %>`, which is undesirable as shown above.
```

Comment:
Can we add a link to `.escape()` and show a simple code example of its usage. I say this because it's all new to our learners.

Reply:
How might we want this to look in terms of usage for the code snippet? @MaoShizhong made a good point in that it's not really best practice to encourage escaping server-side. I can't really think of an instance that `escape()` should or would be used.

Reply:
Perhaps something like the following (very rough and wordy suggestion):

> So why not escape the data when we receive it by adding `.escape()` to the end of our `body()` validation chain instead? "Dangerous characters" are only dangerous at the point of use, and also only in certain contexts. What's "dangerous" for HTML may not be dangerous for SQL and vice versa, and they won't pose risks until they get used in those contexts.
> 
> Also, if we have data with HTML entities after escaping them, if we used escaped output with (`<%= %>`), then `&lt;` wouldn't become `<` but output literally as the text `&lt;`. We'd need to either unescape them then re-escape them via escaped output, or end up using unescaped output (`<%- %>`), which, like using `.innerHTML`, is undesirable as shown above.